File: solutions/Python/merge_intervals_sort_and_merge.py
from typing import List

def merge(intervals: List[List[int]]) -> List[List[int]]:
    """Sort and merge approach: O(n log n) time, O(n) space"""
    if not intervals:
        return []
    intervals.sort()
    merged = [intervals[0]]
    for start, end in intervals[1:]:
        if start <= merged[-1][1]:
            merged[-1][1] = max(merged[-1][1], end)
        else:
            merged.append([start, end])
    return merged

print(merge([[1, 3], [2, 6], [8, 10], [15, 18]]))

File: solutions/Java/merge_intervals_sort_and_merge.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length == 0) return new int[0][0];

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        List<int[]> merged = new ArrayList<>();
        int[] current = intervals[0];

        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] <= current[1]) {
                current[1] = Math.max(current[1], intervals[i][1]);
            } else {
                merged.add(current);
                current = intervals[i];
            }
        }
        merged.add(current);

        return merged.toArray(new int[0][]);
    }
}

int[][] result = new Solution().merge(new int[][]{{1, 3}, {2, 6}, {8, 10}, {15, 18}});
for (int[] interval : result) {
    System.out.println(Arrays.toString(interval));
}
